Updated the porting guide (#427)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/porting-unreal-project-to-gdk.md

<%(Callout type="warn" message="This [pre-alpha](https://docs.improbable.io/reference/latest/shared/release-policy#maturity-stages) release of the SpatialOS Unreal GDK is for evaluation and feedback purposes only, with limited documentation - see the guidance on [Recommended use]({{urlRoot}}/index#recommended-use)")%>

# How to port a native Unreal project to the SpatialOS GDK for Unreal

## Prerequisites

<%(TOC)%>

> Follow the [setup guide]({{urlRoot}}/setup-and-installing) to the end of **Setting up the Unreal GDK module and Starter Project** > **Cloning**.

Reviewer: There's a problem here: The cloning steps above talk specifically about cloning into StarterProject. For this doc, we want you to clone into the project you want to port. Separating the starter docs into different pages might help with this, and we can say "only do the engine setup" etc, then show you how to/where to clone here.

cc: @ernopp @jared-improbable

`<YourProject>` - Name of your game's `.uproject` e.g. `StarterProject.uproject`  

## Setting up the project structure
1. Navigate to the root directory of the GDK repository you cloned (this should be `<ProjectRoot>\<GameRoot>\Plugins\UnrealGDK\`) and run `Setup.bat` in a terminal window or by double-clicking the file. This requires authorization with your SpatialOS account via a web-browser.
1. We now require some files and folders for the SpatialOS Unreal GDK to run correctly with your project, we are going to obtain these by copying them from the `StarterProject` repo that you cloned earlier. Open the StarterProject repo and copy all of the files and folders except for the `\Game\` folder into your `<ProjectRoot>` folder. Your folder structure should now resemble:
   
    ``` cpp
    \<ProjectRoot>\.git\
    \<ProjectRoot>\<GameRoot>\
    \<ProjectRoot>\spatial\
    \<ProjectRoot>\.gitignore
    \<ProjectRoot>\LaunchClient.bat 
    \<ProjectRoot>\LaunchServer.bat
    \<ProjectRoot>\LICENSE.md
    \<ProjectRoot>\ProjectPaths.bat
    \<ProjectRoot>\README.md
    etc...
    ```

> 1. Open **` \<ProjectRoot>\ProjectPaths.bat`** for editing and replace `Game` (located at `set PROJECT_PATH=Game`) with your `<GameRoot>` folder name, and `StarterProject` (located at `set GAME_NAME=StarterProject`) with the name of your game's `.uproject` (which we'll refer to as `<YourProject>`).

Author: Nice. Done.

` Our helper scripts require configuration to work correctly. Set up your project paths:`

## Adding the SpatialGDK module to your project

> 1. In your project's `*.build.cs` file, add "SpatialGDK" to the `PublicDependencyModuleNames`.

Author: Technically we need them to add `"SpatialGDK"` so changing it to that.

> <%(Callout type="warn" message="This [pre-alpha](https://docs.improbable.io/reference/latest/shared/release-policy#maturity-stages) release of the SpatialOS Unreal GDK is for evaluation and feedback purposes only, with limited documentation - see the guidance on [Recommended use]({{urlRoot}}/index#recommended-use)")%>

Reviewer 2: Blast this bit away!

> 

Reviewer 2: Note to self - title may have to change. Under discussion in the Doc of Docs but leave for now.

> 

Reviewer 2: I think we need to have the cloning info for the user's game project in here?

Author: What do you mean by the `cloning info` here?
The `## Prerequisites` specifies the user must have reached the cloning section from the set up guide.

Reviewer 2: Ignore this!

> 

Reviewer 2: I still think the steps need numbers - it's super easy to get lost in a lot of steps without numbers. 
And either the steps should be not at the title level 2 (level 3 instead to get rid of the page lines) or the page split up. I think either would make it more digestible.

> # How to port a native Unreal project to the SpatialOS GDK for Unreal

Reviewer 2: This doc is flipping between backslash and forward slash for directories in a machine (not just GitHub or URLs).
Our standard is to use forward slash. 
Is there any reason to not use this for Windows-only (like copy-pasting file paths?)
Unless there is a strong reason not to, I suggest we use forward slash this even when it's just Windows because it makes it look uniform and it makes documenting easier and therefore more efficient (that is because we don't have to look and think "Is this a URL or an internal path?" ).

> Before porting your project, you need to install and set up the SpatialOS GDK: follow the [set up guide]({{urlRoot}}/setup-and-installing) to the end of **Setting up the Unreal GDK module and Starter Project** > **Cloning**. Additionally make sure your Spatial CLI is up to date: from a terminal window, run the command `spatial update`.

Reviewer 3: Maybe change this to bullet points?

```
Before porting your project:
* Install and set up the SpatialOS GDK: follow the [set up guide]({{urlRoot}}/setup-and-installing) to the end of **Setting up the Unreal GDK module and Starter Project** > **Cloning**.
* Make sure your Spatial CLI is up to date: from a terminal window, run the command `spatial update`.

<%(TOC)%>

> Before porting your project, you need to install and set up the SpatialOS GDK: follow the [set up guide]({{urlRoot}}/setup-and-installing) to the end of **Setting up the Unreal GDK module and Starter Project** > **Cloning**.

Reviewer 2: Adding note to self to update this. Please could we have a "TODO: update the setup link" in the comment here.

## Terms used in this guide

> `<ProjectRoot>` - The root folder of your Unreal project.  

Reviewer 4: @ElleEss It feels like these terms will be regularly used through out the docs, maybe they should be in the glossary. I had to use some of them in the Toolbar docs.

Reviewer 2: OK - if we are using them throughout the doc will add to the living Glossary GDoc and tag you as I think these should probably have a special place in the Glossary.

Reviewer 2: Also, @Giray and @Josh want to refer to your "project" as your "game" in instructions (see below), so we need to standardise on this.

Reviewer: You're @'ing random strangers Laura :)

> `<ProjectRoot>` - The root folder of your Unreal project.  

Reviewer 2: Suggest covering all bases:
`<ProjectRoot>` - The root folder of your Unreal game project.  
`<GameRoot>` - The folder containing your game project’s `.uproject` and source folder (for example, `<ProjectRoot>/ShooterGame/`).
`<YourProject>` - Name of your game project’s `.uproject` (for example, `StarterProject.uproject`).

Author: I'll make the changes to `game project's`. We may want to do a final pas after some updates to make sure these all make sense in their context though.

`<GameRoot>` - The folder containing your game's `.uproject` and source folder (for example, `<ProjectRoot>/ShooterGame/`).  
`<YourProject>` - Name of your game's `.uproject` (for example, `StarterProject.uproject`).

> ## Set up the project structure

Reviewer 2: Set up the project structure

->???

 Set up the game project structure

Author: To me this doesn't make much sense since the `game project` is the `GameRoot` in my eyes.
The `project` is the the folder containing the `GameRoot` since that has the spatial elements and whatever else the dev may need.

Reviewer 2: We are flipping between "project" and "game" all over the docs and when you are reading this, the flipping between makes you think "what does this mean?" - it seems a little thing but changing terms are unsettling and confusing when you are in a strange new world and trying to make sense of things. It means you have to read things twice and you don't want to as there's enough to learn already. 
We should be consistent, at least in the same set of instructions (that is in this porting doc) even if we can't do it throughout all the docs. We should use the same term, unless there are occasions where it makes sense to use the different one. 
Because the terms are flipping all over the docs, I am suggesting using "game project" here as a catch-all as it's so inconsistent everywhere, but to be honest, I don't care which we use, as long as we don't keep switching. 
To me, it makes most sense to say "project" everywhere in this doc as we are dealing with files and talking about Unreal projects, apart from where we are talking about launching your game.

### 1. Set up the project structure
1. Your project needs some extra files and folders to run with the GDK; you can copy these from the StarterProject repository that you cloned earlier.

>     To do this: either in a terminal window or your file manager, navigate to the root of the `StarterProject` repository and copy all of the files and folders except the `\Game\` _folder_ into your `<ProjectRoot>` folder. Your game's folder structure should now resemble:

Reviewer 3: How about we explicitly list out the files that need to be copied and only those files?

    
        ``` cpp

>         \<ProjectRoot>\.git\

Reviewer 3: Is this needed? :o

        \<ProjectRoot>\<GameRoot>\
        \<ProjectRoot>\spatial\

>         \<ProjectRoot>\.gitignore

Reviewer 3: This too

>     

Reviewer 2: ..Your game's folder structure 
->
Your game project's folder structure

# Port your Unreal project to the GDK

## Before you start

Before porting your project: 
* Install and set up the SpatialOS GDK: follow the [set up guide]({{urlRoot}}/setup-and-installing) to the end of **Setting up the Unreal GDK module and Starter Project** > **Cloning**. 
* Make sure your Spatial CLI is up to date: from a terminal window, run the command `spatial update`.
<!-- // TODO: Update the set up link when ready -->

### Terms used in this guide
`<ProjectRoot>` - The root folder of your Unreal project.  

> `<GameRoot>` - The folder containing your game project's `.uproject` and source folder (for example, `<ProjectRoot>/ShooterGame/`).  

Reviewer 5: Start with defining `<GameRoot>` as the folder containing the `.uproject`.
Then, if `<GameRoot>` folder lives inside another folder, then that is the `<ProjectRoot>`, otherwise you should create a folder and move `<GameRoot>` inside of it.

`<YourProject>` - Name of your game project's `.uproject` (for example, `StarterProject.uproject`).

## Start porting!

<%(TOC)%>

### 1. Set up the project structure
1. Your game's project needs some extra files and folders to run with the GDK; you can copy these from the StarterProject repository that you cloned earlier.

    To do this: either in a terminal window or your file manager, navigate to the root of the `StarterProject` repository and copy all of the files and folders below:  

    ``` cpp
    \StarterProject\spatial\
    \StarterProject\LaunchClient.bat 
    \StarterProject\LaunchServer.bat
    \StarterProject\ProjectPaths.bat
    ```
    Your game's folder structure should now resemble:

    ``` cpp
    \<ProjectRoot>\<GameRoot>\
    \<ProjectRoot>\spatial\
    \<ProjectRoot>\LaunchClient.bat 
    \<ProjectRoot>\LaunchServer.bat
    \<ProjectRoot>\ProjectPaths.bat
    etc...
    ```

1. Our helper scripts require configuration to work correctly. Set up your project paths:
   

>    Open **`\<ProjectRoot>\ProjectPaths.bat`** for editing and:

Reviewer 6: What sort of errors could users expect if they don't update this .bat file?

Author: Added some extra info

# Port your Unreal project to the GDK

As an experienced UE4 developer, you likely have a prototype or a game already. The GDK allows you to very easily port it over to SpatialOS. This guide allows you to kickstart your SpatialOS journey - by the end of it your game will run on a *single server-worker* on SpatialOS. You will be ready to start adding multiserver logic to take advantage of the distributed architecture of SpatialOS.

## Before you start

Before porting your project:  

* Install and set up the SpatialOS GDK: follow the [set up guide]({{urlRoot}}/content/get-started/introduction) to the end of **Setting up the Unreal GDK module and Starter Project** > **Cloning**.  
* Make sure your Spatial CLI is up to date: from a terminal window, run the command `spatial update`. 
<!-- // TODO: Update the set up link when ready -->

### Terms used in this guide

> `<GameRoot>` - The folder containing your game project's `.uproject` and source folder.  

Reviewer 5: ```suggestion
`<GameRoot>` - The folder containing your game project's `.uproject` and `Source` folder.  
```

`<ProjectRoot>` - The folder containing your `<GameRoot>`.  
`<YourProject>` - Name of your game project's `.uproject` (for example, `\<GameRoot>\StarterProject.uproject`).

## Start porting!

<%(TOC)%>

### 1. Set up the project structure
1. Ensure you have a `<ProjectRoot>`. If your `<GameRoot>` lives inside of a self-contained folder already, this is your `<ProjectRoot>`. If not, you should create a new folder to represent your `<ProjectRoot>` and move your `<GameRoot>` inside of it.  
   
    Your project structure should take the form of `\<ProjectRoot>\<GameRoot>\<YourProject>.uproject`

    For example:
    `\StarterProject\Game\StarterProject.uproject`
    
    > This step is essential as the `spatial` folder must be located in the directory above your `<GameRoot>`. This is so the GDK scripts work correctly with Unreal.

> 1. Your game's project needs some extra files and folders to run with the GDK; you can copy these from the StarterProject repository that you cloned earlier.

Reviewer 5: Link to the #before-you-start?

    To do this: either in a terminal window or your file manager, navigate to the root of the `StarterProject` repository and copy all of the files and folders below to your `<ProjectRoot>`:  

    ``` cpp
    \StarterProject\spatial\
    \StarterProject\LaunchClient.bat 
    \StarterProject\LaunchServer.bat
    \StarterProject\ProjectPaths.bat
    ```
    Your game's folder structure should now resemble:

    ``` cpp
    \<ProjectRoot>\<GameRoot>\
    \<ProjectRoot>\spatial\
    \<ProjectRoot>\LaunchClient.bat 
    \<ProjectRoot>\LaunchServer.bat
    \<ProjectRoot>\ProjectPaths.bat
    etc...
    ```

1. Our helper scripts require configuration to work correctly. Set up your project paths:  
   Open **`\<ProjectRoot>\ProjectPaths.bat`** for editing and:  

    * In `set PROJECT_PATH=Game`, replace `Game` with your `<GameRoot>` folder name.  

>     * In `set GAME_NAME=StarterProject`, replace `StarterProject` with the name of your game's `.uproject` (which we'll refer to as `<YourProject>`).  

Reviewer 5: I think we already defined `<YourProject>` at this point?

> # How to port a native Unreal project to the Unreal GDK

Reviewer 2: # How to port a native Unreal project to the SpatialOS GDK for Unreal

## Prerequisites

<%(TOC)%>

> Follow the [setup guide]({{urlRoot}}/setup-and-installing) to the end of **Setting up the Unreal GDK module and Starter Project** > **Cloning**.

Author: Nice idea. Done.

> > To speed up build times we recommend installing IncrediBuild, FastBuild, or another build distributor.

Reviewer: Let's remove the Incredibuild talk altogether. We no longer add to UE4 build times (except the first engine build), so we don't affect anyone's decision to use Incredibuild.

> ## Glossary

Author: Sure can change it to `Terms used in this guide` :D

`<ProjectRoot>` - The root folder of your Unreal project.  
`<GameRoot>` - The folder containing your game's `.uproject` and source folder. e.g. `<ProjectRoot>/ShooterGame/`  

> `<YourProject>` - Name of your game's `.uproject` e.g. `StarterProject.uproject`  

Author: Thanks for the tip.

## Setting up the project structure
1. Navigate to the root directory of the GDK repository you cloned (this should be `<ProjectRoot>\<GameRoot>\Plugins\UnrealGDK\`) and run `Setup.bat` in a terminal window or by double-clicking the file. This requires authorization with your SpatialOS account via a web-browser.

> 1. We now require some files and folders for the SpatialOS Unreal GDK to run correctly with your project, we are going to obtain these by copying them from the `StarterProject` repo that you cloned earlier. Open the StarterProject repo and copy all of the files and folders except for the `\Game\` folder into your `<ProjectRoot>` folder. Your folder structure should now resemble:

Reviewer: `SpatialOS Unreal GDK` -> `the GDK`

Follow the [setup guide]({{urlRoot}}/setup-and-installing) to the end of **Setting up the Unreal GDK module and Starter Project** > **Cloning**.

> > To speed up build times, install IncrediBuild, FastBuild, or another build distributor.

Reviewer: Maybe we should just recommend Incredibuild. Does anything else work without our help?

> ## Glossary

Reviewer: "This workflow is very subject to change" - we can remove that I think

Author: Removed.

`<ProjectRoot>` - The root folder of your Unreal project.  

> `<GameRoot>` - The folder containing your Unreal game files and source code. Usually called ‘Game’ e.g. `<ProjectRoot>/Game`  

Reviewer: "The folder containing your game module". And remove the "Usually called Game" bit,that's not really true.

Reviewer: `Source/ShooterGame` might be an example that is familiar.

> `<YourProject>` - Name of your games `.uproject` e.g. `StarterProject.uproject`  

Reviewer: "game's"

## Setting up the project structure

> 1. Navigate to the root directory of the SpatialOS Unreal GDK repository you cloned (this should be `<ProjectRoot>\<GameRoot>\Plugins\UnrealGDK\`) and run `Setup.bat` in a terminal window or by double-clicking the file. This requires authorization with your SpatialOS account via a web-browser.

Reviewer: Let's just call it "GDK repository" because we're now called The SpatialOS GDK for Unreal Engine and ain't nobody got time to say that.

2. We now require some files and folders for the SpatialOS Unreal GDK to run correctly with your project, we are going to obtain these by copying them from the `StarterProject` repo that you cloned earlier. Open the StarterProject repo and copy all of the files and folders except for the `\Game\` folder into your `<ProjectRoot>` folder. Your folder structure should now resemble:
```

> \<ProjectRoot>\ci\

Reviewer 4: This folder should not be required for end users, it is just a folder we have in our repos to enable our CI

\<ProjectRoot>\LaunchClient.bat 
etc...
```
1. Open **` \<ProjectRoot>\ProjectPaths.bat`** for editing and replace `Game` (located at `set PROJECT_PATH=Game`) with your `<GameRoot>` folder name, and `StarterProject` (located at `set GAME_NAME=StarterProject`) with the name of your games uproject (which we'll refer to as `<YourProject>`).

## Adding the SpatialGDK module to your project

> 1. In your project's `*.build.cs` file, add "SpatialGDK" to the `PublicDependencyModuleNames`.

Reviewer: The snippet below seems to be borked. 

https://docs.improbable.io/unreal/pre-alpha/content/porting-unreal-project-to-gdk

Author: Added an arbitrary code specifier to try and fix it.

            );
    ```

### 3. Build your project
Set up your Unreal game project to work with the GDK for Unreal fork of the Unreal Engine, which you cloned and installed in the [Before you start]({{URLRoot}}/content/porting-unreal-project-to-gdk#Before-you-start) section. To do this:
<!-- // TODO: before you start relative link  -->
1. In File Explorer, navigate to `<ProjectRoot>\<GameRoot>`.
1. Right-click your `.uproject` file and select **Switch Unreal Engine version**.
1. Select the path to the Unreal Engine fork you cloned earlier.
1. In File Explorer, right-click your `.uproject` file and select **Generate Visual Studio project files**.
1. Open the generated solution file in Visual Studio and in the Development Editor build configuration, compile and run the project.

### 4. Modify Unreal classes for GDK compatibility

> It is necessary to modify your games `GameInstance` class(es) to work properly with the GDK.  

Reviewer 6: Typo: "your games" -> "your game's"

> 1. Make your game's `GameInstance` inherit from `SpatialGameInstance`. 

Reviewer 6: Typo: "inherit" -> "inherits"

>     };

Reviewer 2: NOTE TO SELF, LAURA - I GOT TO HERE

Reviewer 2: back in the saddle.

Author: You're amazing.

Reviewer 2: Thanks, so are you. I didn't get much further than the lines below yesterday. Downed painkillers and back on it now.

    ```

1. In your project's `PlayerController` header file(s), add the public method declaration
`virtual void InitPlayerState() override;`.

> 1. In your game's `PlayerController` class(es), add the following definition of the `InitPlayerState` function:

Reviewer 2: Is there a reason we switch to "game" here? It's clearer to users to use consistent terms unless switching to game is more meaningful.

Author: Oh I see what you mean. Discussed with Giray and we're switching them all to `game's` since this is more accurate in this case.

Reviewer 2: OK. In the context of the whole document, it uses "project" - I am following the whole document at the same time as this PR as we need to look at edits in context.
So we need to use a consistent term, unless it makes sense to flip terms. If we use different terms unnecessarily, it throws people off and makes them have to think about what they are doing and this makes the process seem hard and complex - we want it to seem smooth and easy!

Author: Here the `game's` makes sense. This `PlayerController` is a part of the `game` while the `game` is a member of the project as a whole.

Reviewer 2: OK, if it makes sense.

1. Specify Singleton Actors  
The GDK uses [Singleton Actors]({{urlRoot}}/content/singleton-actors) - these are server-side authoritative Actors which are a single source of truth for both operations and data across a multi-server simulation; `GameState` and `GameMode` are examples of this. To tell SpatialOS how to replicate Singleton Actors, the GDK  has a `UCLASS` [specifier (Unreal docs)](https://docs.unrealengine.com/en-US/Programming/UnrealArchitecture/Reference/Classes/Specifiers). You add this to classes you want to be Singleton Actors. 
    > If you have not yet created a `GameState` or `GameMode` for your game and are still using the defaults, you must either create Blueprints or native `GameState` and `GameMode` class(es) now.  

    1. If your game's `GameState` is a `.cpp` file, locate it and mark it as a Public Singleton by modifying the `UCLASS` specifier as shown:
        ```
        UCLASS(SpatialType=Singleton)
        ```

>         If your game's `GameState` is a Blueprint, you need to open and edit it in the Blueprint Editor: from the Blueprint Editor toolbar, navigate to the **Class Settings**. In **Class Options**, click the advanced drop down and check **Spatial Type**, in the Spatial Description text box enter `Singleton`.

Reviewer 6: Make this more obvious: "click the advanced drop down" -> "click the 'Advanced' drop down"

    > These lines ensure that the SpatialOS Unreal GDK can override Unreal's network replication.
    >

>     > If you ever need to run with default Unreal networking for workflow or validation reasons, untick `Spatial Networking` in the `Play` drop down menu in the Unreal editor. This flag is maintained in your Unreal config files under `/Script/EngineSettings.GeneralProjectSettings` and is also valid for non-editor builds.

Reviewer 4: Is it worth adding the name of the config file where this entry is written to?

Author: Added.

    >

>     > You can easily switch between networking modes using this checkbox. As the SpatialOS Unreal GDK is in pre-alpha, switching back to Unreal's networking may be a viable way to work around some issues you may encounter, and may speed up your iteration. However, in Unreal networking mode you lose access to key multi-server features of the GDK, which may lead to erratic behavior.

Reviewer 7: `Is in alpha`? We should probably do a full scan for `per-alpha`

Reviewer: I have a pre-release checklist item to do that.

Author: Line removed.

>     > You can easily switch between networking modes using this checkbox. As the SpatialOS Unreal GDK is in pre-alpha, switching back to Unreal's networking may be a viable way to work around some issues you may encounter, and may speed up your iteration. However, in Unreal networking mode you lose access to key multi-server features of the GDK, which may lead to erratic behavior.

Reviewer 4: `As the SpatialOS Unreal GDK is in pre-alpha` > `As the SpatialOS Unreal GDK is in alpha`

Reviewer 4: Maybe we should change this to reflect its debugging purpose rather than as a means to avoid missing features in the GDK

Author: Done.

    >

>     > In the future, we expect to achieve parity with Unreal networking in terms of iteration speed and stability, thus removing the need for a networking mode switch.

Reviewer: Remove this line.

> 2. To provide context to SpatialOS on how to replicate singleton classes (such as GameState and GameMode) we have added a `UCLASS` macro which provides said context. The next steps detail how to do this. 

Reviewer 7: We have a doc section on singletons, link to it here?

Author: Good idea. done.

> 2. To provide context to SpatialOS on how to replicate singleton classes (such as GameState and GameMode) we have added a `UCLASS` macro which provides said context. The next steps detail how to do this. 

Reviewer 4: `2.` ->`1.`

> 2. To provide context to SpatialOS on how to replicate singleton classes (such as GameState and GameMode) we have added a `UCLASS` macro which provides said context. The next steps detail how to do this. 

Reviewer 4: We should link to the singleton class page in the docs.

`UCLASS` macro  -> `UCLASS` specifier and add a link (https://docs.unrealengine.com/en-US/Programming/UnrealArchitecture/Reference/Classes/Specifiers)

Author: Cheers, added doc link too.

> 3. If your games '`GameState`' is a cpp file, locate it and mark it as a public singleton by modifying the UCLASS macro as shown:

Reviewer 4: `3.` ->`1.`

    ```
    UCLASS(SpatialType=Singleton)
    ```
    If your games '`GameState`' is a blueprint then you will need to open the full blueprint editor for that GameState and navigate to the `Class Settings` in the blueprint editor toolbar. In the `Class Options` tick the `Spatial Type` checkbox and add the tag `Singleton` to the `Spatial Description` textbox.

> 4. If your games '`GameMode`' is a cpp file, locate it and mark it as a private singleton by modifying the UCLASS macro as shown:

Reviewer 4: Uclass specifier

    ```
    UCLASS(SpatialType=(Singleton,ServerOnly))
    ```
    If your games '`GameMode`' is a blueprint then you will need to open the full blueprint editor for that GameMode and navigate to the `Class Settings` in the blueprint editor toolbar. In the `Class Options` tick the `Spatial Type` checkbox and add the tag `Singleton,ServerOnly` to the `Spatial Description` textbox.

> 5. Marking these singleton classes as `Spatial Type` will cause spatial schema to be generated for said classes when pressing the `Schema` button in the `SpatialOSGDKToolbar`. We now need to mark the generated schema components as streaming queries for your UnrealWorker worker configuration, this ensures your UnrealWorkers always have these singletons checked out. The generated schema is located at `<ProjectRoot>\spatial\schema\improbable\unreal\generated`.

Reviewer 7: Is it worth mentioning that this fiddly step will become obsolete in the near future?

Author: We'll update the docs when it's done I believe.

>     Open the UnrealWorker worker config located in `<ProjectRoot>\spatial\workers\unreal\spatialos.UnrealWorker.worker.json` and add your new Singleton generated component(s) to the streaming queries.

Reviewer: Link to SpatialOS streaming queries documentation

    
    For example:

            "streaming_query": [
            {
                "global_component_streaming_query": {
                "component_name": "improbable.unreal.GlobalStateManager"
                }
            },
            {
                "global_component_streaming_query": {
                "component_name": "improbable.unreal.generated.yourprojectgamestate.YourProjectGameStateMultiClientRepData"
                }
            },
            {
                "global_component_streaming_query": {
                "component_name": "improbable.unreal.generated.yourprojectgamemode.YourProjectGameModeMultiClientRepData"
                }
            }
            ],
    
## Generating code for SpatialOS C++ workers and building your project
Set your Unreal project to work with the SpatialOS GDK for Unreal fork of the Unreal Engine, which you cloned and installed earlier. To do this:

> 1. In File Explorer, navigate to `<ProjectRoot>\<GameRoot>`.

Reviewer 7: 1. 1. 1. 1.

> 2. Right-click your uproject file and select **Switch Unreal Engine version**.

Reviewer 4: `1.` for each entry in the list

Reviewer 4: .uproject file

3. Select the path to the Unreal Engine fork you cloned earlier.

> 4. In a terminal window, from `<ProjectRoot>\<GameRoot>\Plugins\UnrealGDK\SpatialGDK\Build\Scripts`, run `BuildWorker.bat <YourProject>Editor Win64 Development <YourProject>.uproject`   

Reviewer: Mention this step is optional, for cloud deployments only.

Reviewer 7: This is for running locally managed workers, not cloud deployments.

Author: Mike is right, this is for the `Editor` config and so won't work in cloud. In our setup guide we include scripts for how to build cloud workers for client and server. Should we include those scripts here too?
Updated to 
```
This allows you to run locally managed workers. For example running a local spatial deployment with `one_worker` or `two_worker` launch configurations.
```

For Example "`BuildWorker.bat StarterProjectEditor Win64 Development StarterProject.uproject`"  
This builds and packages your workers ready for a Spatial deployment.
5. In File Explorer, right-click your uproject file and select **Generate Visual Studio project files**.
6. Open the solution in Visual Studio, and compile and run the project in the Development Editor configuration.

## Running the Schema and Snapshot Generator
1. In the Unreal Editor, on the SpatialOS Unreal GDK toolbar, click  the **Schema** button to run the `Schema Generator`.
2. On the same toolbar, click the **Snapshot** button.

## Running your game

> 1. In the Unreal Editor, on the toolbar, open the **Play** drop-down menu and check the box next to **Run Dedicated Server** as well as the checkbox for **Spatial Networking**

Reviewer: Can you add screenshots here? This is arguably our most critical document.

Reviewer 7: We should modify the engine to default dedicated server to on. It's an easy step to miss.

Author: Yeah let's include screenshots if possible.

> 2. On the SpatialOS Unreal GDK toolbar, click the green **Launch** (not the default Launch button in the Unreal Editor), `spatial local launch` will be ran in a seperate console window, wait for the output `SpatialOS ready`.

Reviewer 7: This also builds the worker configs.

>     > These lines ensure that the SpatialOS Unreal GDK can override Unreal's network replication.

Reviewer 2: Can't leave a comment on them as they are in the original doc but the titles with gerunds ("ing" verbs used as nouns) - we should amend to imperatives (run, modify, generate, add).

Reviewer 2: Also this title "Adding Unreal GDK configurations" needs to be:

"Adding GDK configurations" 

(We don't say "Unreal GDK" now.)

>     > These lines ensure that the SpatialOS Unreal GDK can override Unreal's network replication.

Reviewer 2: I think we need background info under the header about the huge new concepts we are casually mentioning in a couple of sentences in step 5. These are:
*  "check out" 
*  "workers" 
*  "schema"
*  "streaming queries"
* "Singleton Actor"

Could we add them to the GDK glossary (with a link through to SPOS docs) and pre-warn users they are going to come up.
Suggestion below.

-----
## Add GDK configurations
The steps below reference the following SpatialOS concepts: </br>
[workers] ({{URLRoot}}/content/glossary#workers) [schema]({{URLRoot}}/content/glossary#schema) [checking out]({{URLRoot}}/content/glossary#check-out] [streaming queries]({{URLRoot}}/content/glossary#streaming-queries) [{{URLRoot}}/content/glossary#singleton-actor]
You don't need to know about them in detail to complete the port.

Author: Sound idea. Added this.

>     >

Reviewer 2: I can't comment on it as it's not an edited part of the page but the line:
"These lines ensure that the SpatialOS Unreal GDK can override Unreal's network replication."
-> 
"These lines ensure that the GDK can override Unreal's network replication."

>     > If you ever need to run with default Unreal networking for workflow or validation reasons, un-tick `Spatial Networking` in the `Play` drop down menu in the Unreal editor. This flag is maintained in your Unreal config file `<GameRoot>/Config/DefaultGame.ini` under `/Script/EngineSettings.GeneralProjectSettings` and is also valid for non-editor builds.

Reviewer 2: Two edits here:

_____

“..or validation reasons, un-tick `Spatial Networking` in the `Play` drop down menu in the Unreal editor."
(this is written backwards to how you instruct someone, giving the specifics first. It also  uses the filename/path formatting rather than the UI formatting and uses "tick" which is British English (sorry, Josh, has to go :-( !!). )
-> 
..or validation reasons, you can switch networking in the Unreal Editor: from the Editor menu, select *Play* and from the drop-down menu, un-check `Spatial Networking` .

----
"This flag is maintained in your Unreal config file `<GameRoot>/Config/DefaultGame.ini` under `/Script/EngineSettings.GeneralProjectSettings` and is also valid for non-editor builds."
-  Slipped into passive voice (who or what maintains it? Do I need to do something?)
 - "Non-Editor" builds are command-line builds, I think?, so let's say what they are as it's helpful to be specific about the use (unless there's loads of other ways to build and they are too numerous to specify).
- Let's standardize the slashes.
- It's a setting in the Editor, I think, so we should call it that.

->
"This setting is valid for Editor and command-line builds. It is stored in your game project's Unreal config file; `<GameRoot>\Config\DefaultGame.ini` under `\Script\EngineSettings.GeneralProjectSettings`."

Author: That looks pretty good. Making the change.

Reviewer 2: Note the first sentence is also back to front and not  so clear. I have suggested:
"..or validation reasons, you can switch networking in the Unreal Editor: from the Editor menu, select Play and from the drop-down menu un-check Spatial Networking ."

Author: Done so.

>     > If you ever need to run with default Unreal networking for workflow or validation reasons, un-tick `Spatial Networking` in the `Play` drop down menu in the Unreal editor. This flag is maintained in your Unreal config file `<GameRoot>/Config/DefaultGame.ini` under `/Script/EngineSettings.GeneralProjectSettings` and is also valid for non-editor builds.

Reviewer 2: In the list item - "In <GameRoot>\Config, open DefaultEngine.ini and add:"
(I can't get to it because it's not a diff but the edits further down mean we need to edit this bit....)
Make:
" 1. In <GameRoot>\Config, open DefaultEngine.ini and add:"
->
" 1. Set up SpatialOS networking. <br/>
In <GameRoot>\Config, open DefaultEngine.ini and add:"

----
This makes the step below make sense.

    >

>     > You can easily switch between networking modes using this checkbox. As the SpatialOS Unreal GDK is in alpha, switching back to Unreal's networking may be a viable way for debugging any issues you may be having and may speed up your iteration times. However, in Unreal default networking mode you lose access to the multi-server features of the GDK, which may lead to erratic behavior.

Reviewer 4: Do we want a screenshot for the check box?

Reviewer 2: I have suggested standarising the menu instructions to the way we always write them, which should make it clearer.

We could link to the toolbar page - although I note that this is not in the toolbar page. 
https://docs.improbable.io/unreal/pre-alpha/content/toolbar

Reviewer 2: Suggest some edits:

"You can easily switch between networking modes using this checkbox. As the SpatialOS Unreal GDK is in alpha, switching back to Unreal's networking may be a viable way for debugging any issues you may be having and may speed up your iteration times. However, in Unreal default networking mode you lose access to the multi-server features of the GDK, which may lead to erratic behavior."

* "You can easily switch between networking modes using this checkbox." <- don't need this - it's clear from how we have written it above.
* SpatialOS Unreal GDK -> the GDK
* may be a viable way -> "useful" probs better as this suggests that SpatialOS networking may be not viable and we don't want to suggest that, I think. :-) 
* any issues you may be having and may speed up your iteration times -> complex use of present continuous passive which we don't need. "iteration" on its own means going round in circles.
* erratic behavior probs needs more attention - make this into a warning? 
* we use two different terms for Unreal's default networking - let's use a consistent term to be super clear and avoid any mental trips.
->
**Warning: **
As the GDK is in alpha, switching back to Unreal default networking mode can be a useful way to debug and so speed up your development iteration. However, you lose access to the multi-server features of the GDK in Unreal default networking mode which may lead to erratic behavior.

Author: Added.

> 

Reviewer 2: This uses the wrong formatting and is slightly back to front. Needs to be:
1. In the Unreal Editor, open the game project's `.uproject` file from the Editor menu, navigate to *Edit* > *Project Settings...* > *SpatialOS Unreal GDK* > *Toolbar*.

> 1. To provide context to SpatialOS on how to replicate [singleton classes]({{urlRoot}}/content/singleton-actors) (such as GameState and GameMode) we have added a `UCLASS` [specifier](https://docs.unrealengine.com/en-US/Programming/UnrealArchitecture/Reference/Classes/Specifiers) which provides said context. The next steps detail how to do this.

Reviewer 2: * We call them Singleton Actors elsewhere, so should be consistent for clarity - unless there is a reason not to?
* I think we might need to qualify what these are a little more because this is likely to be the first time a user comes across the term. 
* The text has also slipped into "we".
* "said context" is awesome but not very doc-ish. 
* We don't need to say "next steps" but make the next steps indented to go under this one.
->
1. Specify Singleton Actors
The GDK uses [Singleton Actors]({{urlRoot}}/content/singleton-actors) - these are server-side authoritative Actors which are a single source of truth for both operations and data across a multi-server simulation; `GameState` and `GameMode` are examples of this. To provide context to SpatialOS on how to replicate Singleton Actors, the GDK  has a `UCLASS` [specifier (Unreal docs)](https://docs.unrealengine.com/en-US/Programming/UnrealArchitecture/Reference/Classes/Specifiers) to provide the context.

Author: Very nice. Added

> 1. If your game's '`GameState`' is a cpp file, locate it and mark it as a public singleton by modifying the UCLASS specifier as shown:

Reviewer 2: Indent me!

Reviewer 2: "cpp" file  -> `.cpp` file

public singleton -> Public Singleton

Reviewer 2: Don't need inverted commas for "GameState" - backslashes are enough.

Reviewer 2: UCLASS - >`UCLASS`

    ```
    UCLASS(SpatialType=Singleton)
    ```

>     If your game's '`GameState`' is a blueprint then you will need to open the full blueprint editor for that GameState and navigate to the `Class Settings` in the blueprint editor toolbar. In the `Class Options` tick the `Spatial Type` checkbox and add the tag `Singleton` to the `Spatial Description` textbox.

Reviewer 2: No need for inverted commas round "GameState".
blueprint -> Blueprint
----
"...blueprint then you will need to open the full blueprint editor for that GameState and navigate to the `Class Settings` in the blueprint editor toolbar. In the `Class Options` tick the `Spatial Type` checkbox and add the tag `Singleton` to the `Spatial Description` textbox."
->
"...Blueprint, you need to open and edit it in the Blueprint Editor: from the Blueprint Editor toolbar, navigate to the **Class Settings** . In **Class Options** check **Spatial Type`** and in the **Spatial Description** text box enter the tag `Singleton` ."

Author: Done.

> 1. If your game's '`GameMode`' is a cpp file, locate it and mark it as a private singleton by modifying the UCLASS specifier as shown:

Reviewer 2: Indent me!

Reviewer 2: 1. If your game's `GameMode` is a `.cpp' file, locate it and mark it as a Private Singleton by modifying the `UCLASS `specifier as shown:

    ```
    UCLASS(SpatialType=(Singleton,ServerOnly))
    ```

>     If your game's '`GameMode`' is a blueprint then you will need to open the full blueprint editor for that GameMode and navigate to the `Class Settings` in the blueprint editor toolbar. In the `Class Options` tick the `Spatial Type` checkbox and add the tag `Singleton,ServerOnly` to the `Spatial Description` textbox.

Reviewer 2: Note Blueprint is a proper noun.

>     If your game's '`GameMode`' is a blueprint then you will need to open the full blueprint editor for that GameMode and navigate to the `Class Settings` in the blueprint editor toolbar. In the `Class Options` tick the `Spatial Type` checkbox and add the tag `Singleton,ServerOnly` to the `Spatial Description` textbox.

Author: Ooo, so it is.

> 1. Marking these singleton classes as `Spatial Type` will cause spatial schema to be generated for said classes when pressing the `Schema` button in the `SpatialOSGDKToolbar`. We now need to mark the generated schema components as [streaming queries](https://docs.improbable.io/reference/13.3/shared/worker-configuration/bridge-config#streaming-queries) for your UnrealWorker worker configuration, this ensures your UnrealWorkers always have these singletons checked out. The generated schema is located at `<ProjectRoot>\spatial\schema\improbable\unreal\generated`.

Reviewer 2: Indent me!

---
* Move into active voice
* Avoid "said"
* Instruction rather than invitation ("you do" rather than "we do"
* Use "server-worker" and "client-worker" because that's what the SpatialOS docs call it.
* This goes into "check out" 
* This goes into "workers" 
* This goes into "schema"
* This goes into "streaming queries"
4 huge new concepts in one sentence - this is a mind blow situation for someone porting a game to SpatialOS for the first time. I think we need to add info about these concepts at the top of this section.
->

"1. Marking these classes as `Spatial Type` enables them to work with SpatialOS the GDK generates  SpatialOS schema for them when you select the **Schema** button in the SpatialOS toolbar.
The resulting generated schema components need to be marked as streaming queries for your server-worker and client-worker (that is `UnrealWorker`) configuration. This ensures your server-workers and client-workers always have these components checked out. The generated schema is located at `<ProjectRoot>\spatial\schema\improbable\unreal\generated`."

>     Open the UnrealWorker worker config located in `<ProjectRoot>\spatial\workers\unreal\spatialos.UnrealWorker.worker.json` and add your new Singleton generated component(s) to the streaming queries.

Reviewer 7: We also need to add the public singletons and the GSM to the UnrealClient.worker.json.

>     Open the UnrealWorker worker config located in `<ProjectRoot>\spatial\workers\unreal\spatialos.UnrealWorker.worker.json` and add your new Singleton generated component(s) to the streaming queries.

Reviewer 2: Have re-ordered the text - it's not clear where I get my component name. Is it from the generated schema files? 

"In your code editor, open the `UnrealWorker.worker.json` configuration file located in `<ProjectRoot>\spatial\workers\unreal\spatialos.UnrealWorker.worker.json` and add your new the newly-generated component(s) to the streaming queries."

Author: With a code update that went in yesterday, we no longer need to specify streaming queries so this part will be removed.

>     For example:

Reviewer 2: For example:
->
The `"streaming_query"` section of your configuration file should look something like this:

>             "streaming_query": [

Reviewer 4: wrap in a ```json

            {
                "global_component_streaming_query": {
                "component_name": "improbable.unreal.GlobalStateManager"
                }
            },
            {
                "global_component_streaming_query": {
                "component_name": "improbable.unreal.generated.yourprojectgamestate.YourProjectGameStateMultiClientRepData"
                }
            },
            {
                "global_component_streaming_query": {
                "component_name": "improbable.unreal.generated.yourprojectgamemode.YourProjectGameModeMultiClientRepData"
                }
            }
            ],
    

> ## Generating code for SpatialOS C++ workers and building your project

Reviewer 2: Generate
build

Reviewer 2: Avoiding the "ing" gerunds and making them more instructions.

Set your Unreal project to work with the SpatialOS GDK for Unreal fork of the Unreal Engine, which you cloned and installed earlier. To do this:
1. In File Explorer, navigate to `<ProjectRoot>\<GameRoot>`.
1. Right-click your `.uproject` file and select **Switch Unreal Engine version**.
1. Select the path to the Unreal Engine fork you cloned earlier.

> 1. In a terminal window, from `<ProjectRoot>\<GameRoot>\Plugins\UnrealGDK\SpatialGDK\Build\Scripts`, run `BuildWorker.bat <YourProject>Editor Win64 Development <YourProject>.uproject`   

Reviewer 2: 1. In a terminal window, from `<ProjectRoot>\<GameRoot>\Plugins\UnrealGDK\SpatialGDK\Build\Scripts`, run `BuildWorker.bat <YourProject>Editor Win64 Development <YourProject>.uproject` 
->
-> Say what we are doing first because we talk about it later. Add full stop.
->
1. Build workers for a local deployment of your game. To do this:<br/>
 In a terminal window, from `<ProjectRoot>\<GameRoot>\Plugins\UnrealGDK\SpatialGDK\Build\Scripts`, run `BuildWorker.bat <YourProject>Editor Win64 Development <YourProject>.uproject` .

> For Example "`BuildWorker.bat StarterProjectEditor Win64 Development StarterProject.uproject`"  

Reviewer 2: " For Example "`BuildWorker.bat StarterProjectEditor Win64 Development StarterProject.uproject`"  "
-> 
Just need to make this clear what we are saying as I has to think about it when I read it, (and no need for inverted commas) so...
->
If, for example, your project is called "StarterProject", this command would be: BuildWorker.bat StarterProjectEditor Win64 Development StarterProject.uproject`

> This allows you to run locally managed workers. For example running a local spatial deployment with `one_worker` or `two_worker` launch configurations.

Reviewer 2: We are introducing "locally managed workers" which doesn't mean anything to users. Have suggested adding something to the beginning of this section.
Original:
"This allows you to run locally managed workers. For example running a local spatial deployment with `one_worker` or `two_worker` launch configurations."
->
->This doesn't mean much to me as I port my project - why would I want to do this? I assume this means using two local server-workers. Also a "launch configuration" doesn't mean much to me. Also we say "server-workers" rather than managed workers.
-> 
Revised (taking into account the additional text at the top of this section):
This means you can now run a local deployment with more than one server; your [launch configuration]({{URLRoot}}/content/glossary#launch-configuration) can have two servers-workers.

1. In File Explorer, right-click your `.uproject` file and select **Generate Visual Studio project files**.

> 1. Open the solution in Visual Studio, and compile and run the project in the Development Editor configuration.

Reviewer 2: What's the solution called? 
->
Revised:
"1. Open the `Name-of-files` solution in Visual Studio and in the Development Editor, compile and run the project."

Author: Tricky one since the solution file is automatically generated based on the `.uporject`.
We could be specific and say `<YourProject>.sln` file....

Perhaps we could skip this all together by saying `Open the generated solution file in Visual Studio and in the Development Editor, compile and run the project.` ?

> ## Running the Schema and Snapshot Generator

Reviewer 2: ## Run the Schema and snapshot generator

> ## Running the Schema and Snapshot Generator

Reviewer 2: The text under this title needs editing (it's not in the diff so having to note it here).

Original:
"In the Unreal Editor, on the SpatialOS Unreal GDK toolbar, click the Schema button to run the Schema Generator."
->
-> Note that we have already generated schema for the singleton actor step - do we defo have to do it again?
->
Revised
" You need to generate schema again and generate a [snapshot]({{URLRoot}}/content/glossary#snapshot) to get your game's deployment started. To do this:<br/>
In the Unreal Editor, on the GDK toolbar, select the **Schema** button to run the [Schema Generator]({{URLRoot}}/content/glossary#schema-generator)."

> 1. In the Unreal Editor, on the SpatialOS Unreal GDK toolbar, click  the **Schema** button to run the `Schema Generator`.

Reviewer 4: Link to the toolbar docs

Reviewer 2: Have suggested link to the Schema Generator definition in the glossary which links to the toolbar doc.

1. On the same toolbar, click the **Snapshot** button.

> ## Running your game

Reviewer 2: We tend to say "launch".
Revised:
"## Launch your game"

1. In the Unreal Editor, on the toolbar, open the **Play** drop-down menu and check the box next to **Run Dedicated Server** as well as the checkbox for **Spatial Networking**

> 1. On the SpatialOS Unreal GDK toolbar, click the green **Launch** (not the default Launch button in the Unreal Editor), this will build your worker configurations (`spatial build build-config`) and `spatial local launch` will be ran in a separate console window, wait for the output "`SpatialOS ready`".

Reviewer 2: Orginal:
"1. On the SpatialOS Unreal GDK toolbar, click the green **Launch** (not the default Launch button in the Unreal Editor), this will build your worker configurations (`spatial build build-config`) and `spatial local launch` will be ran in a separate console window, wait for the output "`SpatialOS ready`"."
->
->The `spatial build` and `spatial local launch` don't really 
->
Revised:
"1. Still in the Unreal Editor but this time from the SpatialOS GDK toolbar, select the green **Launch**  button (not the default Launch button from the Unreal Editor toolbar). This builds your [worker configuration]({{URLRoot/content/glossary#worker-configuration) file and launches your game in a local deployment. <br/>
**Launch** opens up a terminal window and runs two SpatialOS command line interface ([CLI]({{URLRoot/content/glossary#cli)) commands: `spatial build build-config` and `spatial local launch`. It is finished when you see `SpatialOS ready` in the terminal window.

> 1. On the main Unreal toolbar, click **Play**.

Reviewer 2: Revised:
"1. Go back to the main Unreal Editor toolbar (not the GDK toolbar) and select **Play**.

> 

Reviewer 2: Add:
"Job done! You have ported your Unreal game to run on SpatialOS."

## Modifying default behavior

> If you wish to modify the default behavior for the SpatialOS GDK for Unreal; such as the default snapshot save path and filename, the spatial launch configuration used when pressing **Launch** and more... you can modify these settings from the Unreal Editor by navigating to `Edit->Project Settings->SpatialOS Unreal GDK->Toolbar`. 

Reviewer 2: No need to have the first sentence as it repeats the title.

REVISED:

"You can modify some of the GDK settings from the Unreal Editor toolbar at **Edit** > **Project Settings** >* *SpatialOS Unreal GDK** > **Toolbar**.
You can change:
*  the snapshot file's filename and location
*  the launch configuration

    >
    > **Warning:** As the GDK is in alpha, switching back to Unreal default networking mode can be a useful way to debug and so speed up your development iteration. However, you lose access to the multi-server features of the GDK in Unreal default networking mode which may lead to erratic behavior.

1. Specify Singleton Actors  
The GDK uses [Singleton Actors]({{urlRoot}}/content/singleton-actors) - these are server-side authoritative Actors which are a single source of truth for both operations and data across a multi-server simulation; `GameState` and `GameMode` are examples of this. To tell SpatialOS how to replicate Singleton Actors, the GDK  has a `UCLASS` [specifier (Unreal docs)](https://docs.unrealengine.com/en-US/Programming/UnrealArchitecture/Reference/Classes/Specifiers). You add this to classes you want to be Singleton Actors.  

    > If you have not yet created a `GameState` or `GameMode` for your game and are still using the defaults, you must either create Blueprints or native `GameState` and `GameMode` class(es) now.  

>     1. If your game's `GameState` is a `.cpp` file, locate it and mark it as a Public Singleton by modifying the `UCLASS` specifier as shown:

Reviewer: ```suggestion
    1. If your game's `GameState` is a `C++` class, locate its header and mark it as a Public Singleton by modifying the `UCLASS` specifier as shown:
```

        ```
        UCLASS(SpatialType=Singleton)
        ```

>         If your game's `GameState` is a Blueprint, you need to open and edit it in the Blueprint Editor: from the Blueprint Editor toolbar, navigate to the **Class Settings**. In **Class Options**, click the **Advanced** drop down and check **Spatial Type**, in the Spatial Description text box enter `Singleton`.

Reviewer: ```suggestion
        If your game's `GameState` is a Blueprint class, you need to open and edit it in the Blueprint Editor: from the Blueprint Editor toolbar, navigate to the **Class Settings**. In **Class Options**, click the **Advanced** drop down and check **Spatial Type**, in the Spatial Description text box enter `Singleton`.
```

>     1. If your game's `GameMode` is a `.cpp` file, locate it and mark it as a Private Singleton by modifying the `UCLASS` specifier as shown:

Reviewer: ```suggestion
    1. If your game's `GameMode` is a `C++` class, locate its header and mark it as a Private Singleton by modifying the `UCLASS` specifier as shown:
```

        ```
        UCLASS(SpatialType=(Singleton,ServerOnly))
        ```

>        If your game's `GameMode` is a Blueprint, you need to open and edit it in the Blueprint Editor: from the Blueprint Editor toolbar, navigate to the **Class Settings**. In **Class Options**, click the **Advanced** drop down and check **Spatial Type**, in the **Spatial Description** text box enter `Singleton,ServerOnly`.

Reviewer: ```suggestion
       If your game's `GameMode` is a Blueprint class, you need to open and edit it in the Blueprint Editor: from the Blueprint Editor toolbar, navigate to the **Class Settings**. In **Class Options**, click the **Advanced** drop down and check **Spatial Type**, in the **Spatial Description** text box enter `Singleton,ServerOnly`.
```

   Marking these Singleton Actor classes as `Spatial Type` enables them to work with SpatialOS as [schema]({URLRoot}}/content/glossary#schema) will now be generated for them. 

### 6. Generate schema and a snapshot
You need to generate [schema]({URLRoot}}/content/glossary#schema) and generate a [snapshot]({{URLRoot}}/content/glossary#snapshot) to get your game's deployment started. To do this:

> 1. In the Unreal Editor, on the GDK toolbar, click  the **Schema** button to run the [Schema Generator]({{URLRoot}}/content/glossary#schema-generator).

Reviewer: ```suggestion
1. In the Unreal Editor, on the GDK toolbar, click the **Schema** button to run the [Schema Generator]({{URLRoot}}/content/glossary#schema-generator).
```

1. On the same toolbar, click the **Snapshot** button which will generate a snapshot for the map currently open in the editor.

### 7. Launch your game
1. Switch your game project to use the SpatialOS networking. To do this: in the Unreal Editor, from the toolbar, open the **Play** drop-down menu and check two checkboxes:
    * Check the box for **Run Dedicated Server**
    * Check the box for **Spatial Networking**

>     > From this drop-down menu it is possible to increase the number of servers that will be launched. For now leave this at 1.

Reviewer: ```suggestion
    > From this drop-down menu it is possible to increase the number of servers that will be launched. For now leave this at 1. This is because there is currently no multiserver logic in your code. This port will be the baseline for you to start building the multiserver game logic.
```

    NetConnectionClassName="/Script/SpatialGDK.SpatialNetConnection"
    ```

    > These lines ensure that the GDK can override Unreal's network replication.
    >

>     > If you ever need to run with default Unreal networking for workflow or validation reasons, you can switch networking in the Unreal Editor: from the Editor menu, select **Play** and from the drop-down menu un-check `Spatial Networking`. This setting is valid for Editor and command-line builds. It is stored in your game project's Unreal config file; `<GameRoot>\Config\DefaultGame.ini` under `\Script\EngineSettings.GeneralProjectSettings`.

Reviewer 5: ```suggestion
    > If you ever need to run with default Unreal networking for workflow or validation reasons, you can switch networking in the Unreal Editor: from the Editor menu, select **Play** and from the drop-down menu un-check `Spatial Networking`. This setting is valid for Editor and command-line builds. It is stored in your game project's Unreal config file; `<GameRoot>\Config\DefaultGame.ini` under `/Script/EngineSettings.GeneralProjectSettings`.
```

>     > If you ever need to run with default Unreal networking for workflow or validation reasons, you can switch networking in the Unreal Editor: from the Editor menu, select **Play** and from the drop-down menu un-check `Spatial Networking`. This setting is valid for Editor and command-line builds. It is stored in your game project's Unreal config file; `<GameRoot>\Config\DefaultGame.ini` under `\Script\EngineSettings.GeneralProjectSettings`.

Reviewer 5: "select Play and" - I think they will just click the button which will start the game.
Maybe something like "press the drop down arrow next to the **Play** button"?

    > If you have not yet created a `GameState` or `GameMode` for your game and are still using the defaults, you must either create Blueprints or native `GameState` and `GameMode` class(es) now.  

    1. If your game's `GameState` is a `C++` class, locate it's header and mark it as a Public Singleton by modifying the `UCLASS` specifier as shown:
        ```
        UCLASS(SpatialType=Singleton)
        ```
        If your game's `GameState` is a Blueprint class, you need to open and edit it in the Blueprint Editor: from the Blueprint Editor toolbar, navigate to the **Class Settings**. In **Class Options**, click the **Advanced** drop down and check **Spatial Type**, in the Spatial Description text box enter `Singleton`.

    1. If your game's `GameMode` is a `C++` class, locate it's header and mark it as a Private Singleton by modifying the `UCLASS` specifier as shown:
        ```
        UCLASS(SpatialType=(Singleton,ServerOnly))
        ```

>        If your game's `GameMode` is a Blueprint class, you need to open and edit it in the Blueprint Editor: from the Blueprint Editor toolbar, navigate to the **Class Settings**. In **Class Options**, click the **Advanced** drop down and check **Spatial Type**, in the **Spatial Description** text box enter `Singleton,ServerOnly`.

Reviewer 5: I believe it will automatically replace the comma with a space (`Singleton ServerOnly`), or at least if you re-open the blueprint.

**Launch** opens up a terminal window and runs two SpatialOS command line interface ([CLI]({{URLRoot/content/glossary#cli)) commands: `spatial build build-config` and `spatial local launch`. It is finished when you see `SpatialOS ready` in the terminal window.
1. On the main Unreal toolbar, click **Play**. 
1. From the SpatialOS GDK toolbar click **Inspector** which will open a local [SpatialOS inspector](https://docs.improbable.io/reference/13.3/shared/operate/inspector) in your web browser. Here you can see the entities and their components present in your deployment, updates are in real-time.
  
**For running a local deployment with managed workers or a cloud deployment take a look at the [glossary section for deployments]({{URLRoot}}/content/glossary#deployment)**

**Job done!** You have ported your Unreal game to run on SpatialOS. Move around and look at the changes reflected in your inspector.

> 

Reviewer 5: Should we add a link to common issues?
